File: app/src/main/java/com/evernote/android/job/sample/reminder/ReminderEngine.java
package com.evernote.android.job.sample.reminder;

import android.annotation.SuppressLint;
import android.app.Notification;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.support.annotation.Nullable;
import android.support.v4.app.NotificationManagerCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.NotificationCompat;

import com.evernote.android.job.JobManager;
import com.evernote.android.job.sample.MainActivity;
import com.evernote.android.job.sample.R;

import net.vrallev.android.context.AppContext;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Random;
import java.util.Set;

/**
 * @author rwondratschek
 */
public class ReminderEngine {

    private static final String REMINDER_ID = "REMINDER_ID";
    private static final String REMINDERS = "REMINDERS";

    @SuppressLint("StaticFieldLeak")
    private static final ReminderEngine INSTANCE = new ReminderEngine();

    public static ReminderEngine instance() {
        return INSTANCE;
    }

    private final Context mContext;
    private final SharedPreferences mPreferences;

    private final List<Reminder> mReminders;

    private final List<ReminderChangeListener> mListeners;

    private int mReminderId;

    private ReminderEngine() {
        mContext = AppContext.get();
        mPreferences = mContext.getSharedPreferences("reminders", Context.MODE_PRIVATE); // poor-man's storage
        mReminderId = mPreferences.getInt(REMINDER_ID, 0);

        mReminders = new ArrayList<>();
        mListeners = new ArrayList<>();

        Set<String> reminders = mPreferences.getStringSet(REMINDERS, null);
        if (reminders != null) {
            for (String value : reminders) {
                mReminders.add(Reminder.fromString(value));
            }
        }
        Collections.sort(mReminders);
    }

    public void addListener(ReminderChangeListener listener) {
        mListeners.add(listener);
    }

    public void removeListener(ReminderChangeListener listener) {
        mListeners.remove(listener);
    }

    public List<Reminder> getReminders() {
        return Collections.unmodifiableList(mReminders);
    }

    @Nullable
    public Reminder getReminderById(int id) {
        for (Reminder reminder : mReminders) {
            if (reminder.getId() == id) {
                return reminder;
            }
        }
        return null;
    }

    public Reminder createNewReminder(long timestamp) {
        mReminderId++;

        Reminder reminder = new Reminder(mReminderId, timestamp);
        mReminders.add(reminder);
        Collections.sort(mReminders);

        int jobId = ReminderJob.schedule(reminder);
        reminder.setJobId(jobId);

        mPreferences.edit().putInt(REMINDER_ID, mReminderId).apply();
        saveReminders();

        int position = mReminders.indexOf(reminder);
        for (ReminderChangeListener listener : mListeners) {
            listener.onReminderAdded(position, reminder);
        }

        return reminder;
    }

    public void removeReminder(int position) {
        removeReminder(position, true);
    }

    /*package*/ void removeReminder(int position, boolean cancelJob) {
        Reminder reminder = mReminders.remove(position);

        if (cancelJob) {
            JobManager.instance().cancel(reminder.getJobId());
        }

        saveReminders();

        for (ReminderChangeListener listener : mListeners) {
            listener.onReminderRemoved(position, reminder);
        }
    }

    public void showReminder(Reminder reminder) {
        PendingIntent pendingIntent = PendingIntent.getActivity(mContext, 0, new Intent(mContext, MainActivity.class), 0);

        Notification notification = new NotificationCompat.Builder(mContext)
                .setContentTitle("Reminder " + reminder.getId())
                .setContentText("Attend MTC Munich!")
                .setAutoCancel(true)
                .setContentIntent(pendingIntent)
                .setSmallIcon(R.drawable.ic_notification)
                .setShowWhen(true)
                .setColor(ContextCompat.getColor(mContext, R.color.accent))
                .setLocalOnly(true)
                .build();

        NotificationManagerCompat.from(mContext).notify(new Random().nextInt(), notification);
    }

    private void saveReminders() {
        Set<String> reminders = new HashSet<>();
        for (Reminder item : mReminders) {
            reminders.add(item.toPersistableString());
        }

        mPreferences.edit().putStringSet(REMINDERS, reminders).apply();
    }

    public interface ReminderChangeListener {
        void onReminderAdded(int position, Reminder reminder);

        void onReminderRemoved(int position, Reminder reminder);
    }
}

File: app/src/main/java/com/evernote/android/job/sample/reminder/ReminderEngine.kt
package com.evernote.android.job.sample.reminder

import android.annotation.SuppressLint
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.os.Build
import android.support.v4.app.NotificationCompat
import android.support.v4.app.NotificationManagerCompat
import android.support.v4.content.ContextCompat
import com.evernote.android.job.JobManager
import com.evernote.android.job.sample.App
import com.evernote.android.job.sample.MainActivity
import com.evernote.android.job.sample.R
import java.util.Random

/**
 * @author rwondratschek
 */

private const val REMINDER_ID = "REMINDER_ID"
private const val REMINDERS = "REMINDERS"

@SuppressLint("StaticFieldLeak")
object ReminderEngine {

    private val context: Context = App.instance

    private val preferences = context.getSharedPreferences("reminders", Context.MODE_PRIVATE) // poor-man's storage
    private var reminderId = preferences.getInt(REMINDER_ID, 0)
    private val reminders: MutableList<Reminder>

    private val listeners = mutableListOf<ReminderChangeListener>()

    init {
        reminders = preferences.getStringSet(REMINDERS, null)
            ?.map { Reminder.fromString(it) }
            ?.toMutableList() ?: mutableListOf()

        reminders.sort()
    }

    fun addListener(listener: ReminderChangeListener) {
        listeners.add(listener)
    }

    fun removeListener(listener: ReminderChangeListener) {
        listeners.remove(listener)
    }

    fun getReminderById(id: Int): Reminder? = reminders.firstOrNull { it.id == id }
    fun getReminders(): List<Reminder> = reminders.toList()

    fun createNewReminder(timestamp: Long): Reminder {
        reminderId++

        val reminder = Reminder(reminderId, timestamp, 0)
        reminders.add(reminder)
        reminders.sort()

        val jobId = ReminderJob.schedule(reminder)
        reminder.jobId = jobId

        preferences.edit().putInt(REMINDER_ID, reminderId).apply()
        saveReminders()

        val position = reminders.indexOf(reminder)
        listeners.forEach { it.onReminderAdded(position, reminder) }

        return reminder
    }

    fun removeReminder(position: Int) {
        removeReminder(position, true)
    }

    fun removeReminder(position: Int, cancelJob: Boolean) {
        val reminder = reminders.removeAt(position)

        if (cancelJob) {
            JobManager.instance().cancel(reminder.jobId)
        }

        saveReminders()

        listeners.forEach { it.onReminderRemoved(position, reminder) }
    }

    fun showReminder(reminder: Reminder) {
        val pendingIntent = PendingIntent.getActivity(context, 0, Intent(context, MainActivity::class.java), 0)

        val channelId = "reminder"

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(channelId, "Job Sample", NotificationManager.IMPORTANCE_LOW)
            channel.description = "Job sample"
            context.getSystemService(NotificationManager::class.java).createNotificationChannel(channel)
        }

        val notification = NotificationCompat.Builder(context, channelId)
            .setContentTitle("Reminder " + reminder.id)
            .setContentText("Hello Droidcon")
            .setAutoCancel(true)
            .setContentIntent(pendingIntent)
            .setSmallIcon(R.drawable.ic_notification)
            .setShowWhen(true)
            .setColor(ContextCompat.getColor(context, R.color.accent))
            .setLocalOnly(true)
            .build()

        NotificationManagerCompat.from(context).notify(Random().nextInt(), notification)
    }

    private fun saveReminders() {
        preferences.edit().putStringSet(REMINDERS, reminders.map { it.toPersistableString() }.toHashSet()).apply()
    }

    interface ReminderChangeListener {
        fun onReminderAdded(position: Int, reminder: Reminder)

        fun onReminderRemoved(position: Int, reminder: Reminder)
    }
}